### DISCUSSION TITLE
Snapcast-Server and Client on same device

### DISCUSSION THREAD

**Discussion by grooWnad3r3r** [rern/rAudio#812] [Feature Requests]
Hello rern,

thank you for the great job.

For Multiroom it will be a nice to have to enable the snapcast-server and client on the same device.
So we can get in sync.

groo

**Comment by rern** [comment]
Please explain more on how you will use it to sync the same device.

**Comment by grooWnad3r3r** [comment]
Hello rern,
thank you for the fast response


My Setup: 
In livingroom raudio is on a Raspi 3B with Hifiberry-hat --> Amplifier 
In three other rooms are self build active speakers with raspi zero and snapcast-client.

If i play music on raspi3 all three rooms have a delay to the livingroom.

Is it possible to pipe the mpd output (or whatever is used) on the raspi3 only to the snapcast server, enable the snapcast client and pipe the output to alsa ?

So all devices can play in sync

I know this is far away from bit-perfect. ;-)


groo

**Comment by rern** [comment]
To adjust the delay: 
Features > SnapClient > Setting (cog icon)

**Comment by grooWnad3r3r** [comment]
Hello rern,

i am sorry, my explanation was bad.

I think this is the actual state.
In this setup there is a delay from the local raspi to the the clients and the local raspy is not in sync with te remote clients
<pre>
audio player software -> snapfifo -> snapserver -> network -> remote snapclient -> alsa
			          |
			            - > local alsa
</pre>



if server and client enabeld on the local device all can play in sync
<pre>
audio player software -> snapfifo -> snapserver -> network -> remote snapclient -> alsa
					         |
					          - > local socket -> local snapclient -> local alsa 
</pre>

but in this version it is not possible to enable snap server and client at the same time

groo

**Reply by ElliottBridge** [reply]
We could really do with this, when wanting to play audio on the server, you want to run snapclient so the audio remains synced with the other snapclients.

**Comment by rern** [comment]
Please try the UPDATE branch:
- Settings > Addons
- rAudio - long-press Update button > Ok to UPDATE branch

**Reply by ElliottBridge** [reply]
Hi Rern,
Tried the update. Unfortunately, it didn't change anything. You still can't enable both snapserver and snapclient on the same device. Perhaps if a delay option could be added instead, maybe that will work instead? But I'm not too sure about that

**Reply by grooWnad3r3r** [reply]
Hello ElliotBridge,
in the gui it´s possible to switch both to "on".
But snapclient wont run because mpd use the default output (im my case a hifiberry hat)
So snapclient stop with an error.

After a service restart all looks good until you try to play music.
Then the mpd grep the default audio interface and snapclient fail
<pre>
rAudio:~  #  systemctl status snapclient
* snapclient.service - Snapcast client
     Loaded: loaded (/usr/lib/systemd/system/snapclient.service; disabled; vendor preset: disabled)
     Active: active (running) since Mon 2022-03-14 18:42:56 CET; 1min 37s ago
       Docs: man:snapclient(1)
   Main PID: 8997 (snapclient)
      Tasks: 2 (limit: 1062)
        CPU: 821ms
     CGroup: /system.slice/snapclient.service
             `-8997 /usr/bin/snapclient --logsink=system --latency=800

Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
Mar 14 18:44:14 rAudio snapclient[8997]: PCM name: default:CARD=sndrpihifiberry, sample rate: 48000 Hz, channels: 2, buffer time: 80000 us, periods: 4, period time: 20000 us>
Mar 14 18:44:16 rAudio snapclient[8997]: Exception: Not enough frames available, requested frames: 960, available: 634
Mar 14 18:44:16 rAudio snapclient[8997]: Failed to get chunk
Mar 14 18:44:18 rAudio snapclient[8997]: XRUN while waiting for PCM: Broken pipe
</pre>

groo

**Reply by ElliottDyson** [reply]
> Hello ElliotBridge,
> in the gui it´s possible to switch both to "on".
> But snapclient wont run because mpd use the default output (im my case a hifiberry hat)
> So snapclient stop with an error.
> 
> After a service restart all looks good until you try to play music.
> Then the mpd grep the default audio interface and snapclient fail
> <pre>
> rAudio:~  #  systemctl status snapclient
> * snapclient.service - Snapcast client
>      Loaded: loaded (/usr/lib/systemd/system/snapclient.service; disabled; vendor preset: disabled)
>      Active: active (running) since Mon 2022-03-14 18:42:56 CET; 1min 37s ago
>        Docs: man:snapclient(1)
>    Main PID: 8997 (snapclient)
>       Tasks: 2 (limit: 1062)
>         CPU: 821ms
>      CGroup: /system.slice/snapclient.service
>              `-8997 /usr/bin/snapclient --logsink=system --latency=800
> 
> Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
> Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
> Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
> Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
> Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
> Mar 14 18:44:14 rAudio snapclient[8997]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
> Mar 14 18:44:14 rAudio snapclient[8997]: PCM name: default:CARD=sndrpihifiberry, sample rate: 48000 Hz, channels: 2, buffer time: 80000 us, periods: 4, period time: 20000 us>
> Mar 14 18:44:16 rAudio snapclient[8997]: Exception: Not enough frames available, requested frames: 960, available: 634
> Mar 14 18:44:16 rAudio snapclient[8997]: Failed to get chunk
> Mar 14 18:44:18 rAudio snapclient[8997]: XRUN while waiting for PCM: Broken pipe
> </pre>
> 
> groo
> 

Yeah, I'm getting audio output related issues too, in my case it's because it's not recognising alsa loopback as an output device. Btw this is my second GitHub account, this is still the original Elliott 😉

**Comment by grooWnad3r3r** [comment]
Hello rern,

if snapclient and snapserver run on the same device we must ensure that in the mpd.conf is only "Snapcast" the audio output.
The snapclient should take care of the audio output to the default audio device.

But rAudio-1 do some magic and recreate the mpd. config  ;-)

And thank you for the fast response and coding :-)

groo

**Comment by grooWnad3r3r** [comment]
Hello rern and Elliot,

i found a solution for this issue.
It´s  with a little nasty cheat.

First set snapclient latency to 0. (or whatever you need for your hardware)
Then edit /etc/mpd.conf as shown
<pre>
#audio_output {
#       name           "HiFiBerry DAC+"
#       device         "hw:0,0"
#       type           "alsa"
#       auto_resample  "no"
#       auto_format    "no"
#       mixer_type     "hardware"
#       mixer_control  "Digital"
#       mixer_device   "hw:0"
#}
audio_output {
        name           "Snapcast"
        type           "fifo"
        path           "/tmp/snapfifo"
        format         "48000:16:2"
        mixer_type     "software"
}
</pre>
As you see i disable  the hardware audio output.
Snapclient take the default audio device for output

Then (nasty cheat)  do a
<pre>
chattr +i /etc/mpd.conf
</pre>

reboot and you have sync multiroom on your pi with snapcast server/client.

Don´t forget after chattr +i  you must do chattr -i if you want to edit mpd.conf


groo

**Reply by rern** [reply]
With update 20220317:
- SnapSever and SnapClient can be enabled at the same time.
- On SnapSever, tap SnapClient icon to connect/disconnect and `mpd.conf` will be reconfigured automatically.

**Reply by grooWnad3r3r** [reply]
Hello rern,

in the GUI i can enable server and client at the same time.
But if i enable snapclient while server is enabled it did not chage the mpd.conf. (sure i do chattr -i )
Enabling Server works as expected. It adds the /tmp/snapfifo to the mpd.conf.
Enabling client leaves the mpd.conf untouched.
The "hardware Audio section" still remain in the conf.
So snapclient can not use the default audio output
<pre>
Mar 24 12:09:42 rAudio snapclient[4524]: Exception in initAlsa: Can't open default:CARD=sndrpihifiberry, error: Device or resource busy
</pre>

Systemhardware:
Raspi 3b
HiFiBerry DAC+
rAudio 1 • 20220317

groo

**Reply by rern** [reply]
Try:
- The latest image file 
   - Without setting up anything
   - Addons > Update
   - Enable SnapServer and SnapClient
- With on-board audio 3.5mm jack to headphones
   - Add a WebRadio and play 
   - Connect / Disconnect SnapClient and see `mpd.conf`

**Reply by grooWnad3r3r** [reply]
Hello rern,

sorry for the late reply.

I do as you say, fresh new install, addon update, enable both snapserver and client , plug in a headphone, start Radio Paradise.

Snapclient is not starting with the gui switch. (tested with systemctl status snapclient)
The switch change the color but do not work on snapclient.
No change in mpd.conf while switching snapclient on/off
Manual starting snapclient works, but needs editing the mpd.conf.


If you need more testing, tell me, i will do.


groo

**Reply by grooWnad3r3r** [reply]
Hello rern,

i tryed again with the latest image and i am happy to say.
It works :-)

Thank a lot for your work
(are there any donate buttons?)


groo

**Comment by yves2** [comment]
Hello Rern

i tried to use snapclient and snapserver at the same time with the last update
 but when connecting the client i can't get any sound 
do i miss something?

here is  the mpd config:

mpd 0.23.9-2
# cat /etc/mpd.conf
bind_to_address        "0.0.0.0"
db_file                "/srv/http/data/mpd/mpd.db"
music_directory        "/mnt/MPD"
playlist_directory     "/srv/http/data/playlists"
replaygain             "off"
state_file             "/var/lib/mpd/mpdstate"
user                   "mpd"

input {
	plugin         "curl"
}

resampler {
	plugin         "soxr"
	quality        "very high"
}

audio_output {
	name           "Snapcast"
	type           "fifo"
	path           "/tmp/snapfifo"
	format         "48000:16:2"
	mixer_type     "software"
}

mpd.service- Music Player Daemon
     Loaded: loaded (/usr/lib/systemd/system/mpd.service; disabled; preset: disabled)
    Drop-In: /usr/lib/systemd/system/mpd.service.d
             └─00-arch.conf
             /etc/systemd/system/mpd.service.d
             └─override.conf
     Active: active (running) since Sun 2022-11-06 00:14:57 CET; 38min ago
       Docs: man:mpd(1)
             man:mpd.conf(5)
   Main PID: 30674 (mpd)
      Tasks: 7 (limit: 4524)
        CPU: 12.615s
     CGroup: /system.slice/mpd.service
             └─30674 /usr/bin/mpd --systemd

Nov 06 00:14:57 rAudio3 systemd[1]: Starting Music Player Daemon...
Nov 06 00:14:57 rAudio3 mpd[30674]: decoder: Decoder plugin 'wildmidi' is unavailable: configuration file does not exist: /etc/timidity/timidity.cfg
Nov 06 00:14:57 rAudio3 systemd[1]: Started Music Player Daemon.
Nov 06 00:15:48 rAudio3 mpd[30674]: player: played "USB/sda1-usb-Kingston_DataTra/Musique/Nino Ferrer/L'Intégrale des enregistrements studio & live (disque « 1 »)/EP 231114/18 - Il me faudra… Natacha.flac"
Nov 06 00:19:29 rAudio3 mpd[30674]: player: played "USB/sda1-usb-Kingston_DataTra/Musique/Nino Ferrer/L'Intégrale des enregistrements studio & live (disque « 1 »)/EP 231114/19 - Ma vie pour rien.flac"
